verification/src/solidity/verifier.rs

```diff
@@ -1,51 +1,281 @@
+                                type_,
+                            )));
+                        }
+                    }
+                }
+                Ok(_) => {
+                    // if key is not "solc" str we may skip the corresponding value
+                    d.skip()?;
+                }
+                Err(err) if err.is_type_mismatch() => {
+                    // if key is not `str` we may skip the corresponding value
+                    d.skip()?;
+                }
+                Err(err) => return Err(err),
+            }
+        }
+
+        // We require that no elements left in the decoder when
+        // the whole map has been processed. That adds another layer
+        // of assurance that encoded bytes are actually metadata hash.
+        if d.datatype().is_ok() {
+            return Err(Error::custom(ParseMetadataHashError::NonExhausted));
+        }
+
+        let solc = solc.map(bytes::Bytes::copy_from_slice);
+        Ok(MetadataHash { solc })
+    }
+
+    fn nil() -> Option<Self> {
+        Some(Self { solc: None })
+    }
+}
+
 /// Wrapper under `evm.deployedBytecode` from the standard output JSON
 /// (https://docs.soliditylang.org/en/latest/using-the-compiler.html#output-description).
 ///
 /// Provides an interface to retrieve parts the deployed bytecode consists of:
-/// actual bytecode participating in EVM transaction execution and optionally metadata hash.
+/// actual bytecode participating in EVM transaction execution and metadata hash.
 #[derive(Clone, Debug, PartialEq)]
-struct DeployedBytecode {}
+struct DeployedBytecode {
+    /// Bytecode without metadata hash
+    bytecode: bytes::Bytes,
+    /// Metadata hash encoded into bytecode
+    metadata_hash: MetadataHash,
+    /// Raw deployed bytecode bytes
+    bytes: bytes::Bytes,
+}
+
+impl DeployedBytecode {
+    /// Returns deployed bytecode without metadata hash
+    pub fn bytecode(&self) -> bytes::Bytes {
+        self.bytecode.clone()
+    }
+
+    /// Returns a metadata hash
+    pub fn metadata_hash(&self) -> &MetadataHash {
+        &self.metadata_hash
+    }
+
+    /// Returns metadata hash encoded as bytes and concatenated with 2 bytes representing its length
+    pub fn encoded_metadata_hash_with_length(&self) -> bytes::Bytes {
+        let start = self.bytecode.len();
+        let end = self.bytes.len();
+        self.bytes.slice(start..end)
+    }
+}
 
 impl FromStr for DeployedBytecode {
     type Err = InitializationError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        todo!()
+        let bytes = DisplayBytes::from_str(s)
+            .map_err(|_| InitializationError::InvalidDeployedBytecode)?
+            .0;
+
+        DeployedBytecode::try_from(bytes)
+    }
+}
+
+impl TryFrom<bytes::Bytes> for DeployedBytecode {
+    type Error = InitializationError;
+
+    fn try_from(encoded: bytes::Bytes) -> Result<Self, Self::Error> {
```

Comment:
do we really need `bytes::Bytes` here? Isn't just `&[u8]` enough?

Reply:
`bytes::Bytes` allows us to clone the structure for O(1) and easily share different parts which is used in current implementation. Besides that, it provides convenient methods to process underlying bytes

Reply:
you can copy `&[u8]` easily too and there is a split function in slices too

Reply:
But you have to solve ownership and lifetime problems which do not occur when you use `bytes::Bytes`